File: sdk/typescript/src/schema-preflight.ts
export const MAX_OUTPUT_SCHEMA_BYTES = 32 * 1024;
export const MAX_OUTPUT_SCHEMA_DEPTH = 16;
export const MAX_OUTPUT_PATTERN_BYTES = 1024;
export const SCHEMA_PREFLIGHT_CODE = "schema_preflight_failed";

export type SchemaIssueCode =
  | "unsupported_keyword"
  | "invalid_schema"
  | "limit_exceeded";

export interface SchemaIssue {
  code: SchemaIssueCode;
  path: string;
  keyword?: string;
  message: string;
}

const allowedKeywords = new Set([
  "type",
  "title",
  "description",
  "properties",
  "required",
  "additionalProperties",
  "items",
  "enum",
  "pattern",
  "minLength",
  "maxLength",
  "minItems",
  "maxItems",
  "uniqueItems",
  "minimum",
  "maximum",
]);

export function outputSchemaIssue(schema: Record<string, unknown>): SchemaIssue | undefined {
  let compact: string | undefined;
  try {
    compact = JSON.stringify(schema);
  } catch {
    return issue("invalid_schema", "", undefined, "schema must be JSON serializable");
  }
  if (compact === undefined) {
    return issue("invalid_schema", "", undefined, "schema must be JSON serializable");
  }
  if (new TextEncoder().encode(compact).byteLength > MAX_OUTPUT_SCHEMA_BYTES) {
    return issue(
      "limit_exceeded",
      "",
      undefined,
      `compact schema exceeds ${MAX_OUTPUT_SCHEMA_BYTES} bytes`,
    );
  }
  return validateSchemaNode(schema, "", 1, true);
}

function validateSchemaNode(
  node: Record<string, unknown>,
  path: string,
  depth: number,
  root: boolean,
): SchemaIssue | undefined {
  if (depth > MAX_OUTPUT_SCHEMA_DEPTH) {
    return issue(
      "limit_exceeded",
      path,
      undefined,
      `schema exceeds the maximum nesting depth of ${MAX_OUTPUT_SCHEMA_DEPTH}`,
    );
  }
  for (const keyword of Object.keys(node).sort()) {
    if (!allowedKeywords.has(keyword)) {
      return memberIssue(
        "unsupported_keyword",
        path,
        keyword,
        `unsupported schema keyword ${JSON.stringify(keyword)}`,
      );
    }
  }
  const typeName = node.type;
  if (typeof typeName !== "string" || !supportedType(typeName)) {
    return memberIssue(
      "invalid_schema",
      path,
      "type",
      "every schema position requires one supported string type",
    );
  }
  if (root && typeName !== "object") {
    return memberIssue("invalid_schema", path, "type", "schema root type must be object");
  }
  for (const keyword of ["description", "title"] as const) {
    if (keyword in node && typeof node[keyword] !== "string") {
      return memberIssue(
        "invalid_schema",
        path,
        keyword,
        `schema ${keyword} must be a string`,
      );
    }
  }
  for (
    const keyword of ["maxItems", "maxLength", "minItems", "minLength"] as const
  ) {
    if (keyword in node && !nonnegativeInteger(node[keyword])) {
      return memberIssue(
        "invalid_schema",
        path,
        keyword,
        `schema ${keyword} must be a nonnegative integer`,
      );
    }
  }
  for (const keyword of ["maximum", "minimum"] as const) {
    if (keyword in node && !jsonNumber(node[keyword])) {
      return memberIssue(
        "invalid_schema",
        path,
        keyword,
        `schema ${keyword} must be a number`,
      );
    }
  }
  if ("uniqueItems" in node && typeof node.uniqueItems !== "boolean") {
    return memberIssue(
      "invalid_schema",
      path,
      "uniqueItems",
      "schema uniqueItems must be a boolean",
    );
  }
  const boundIssue = validateBoundOrder(node, path);
  if (boundIssue) return boundIssue;
  if (typeName !== "string") {
    const keyword = firstPresent(node, "maxLength", "minLength", "pattern");
    if (keyword) {
      return memberIssue(
        "invalid_schema",
        path,
        keyword,
        "string schema keywords require type string",
      );
    }
  }
  if (typeName !== "array") {
    const keyword = firstPresent(node, "maxItems", "minItems", "uniqueItems");
    if (keyword) {
      return memberIssue(
        "invalid_schema",
        path,
        keyword,
        "array schema bounds require type array",
      );
    }
  }
  if (typeName !== "number" && typeName !== "integer") {
    const keyword = firstPresent(node, "maximum", "minimum");
    if (keyword) {
      return memberIssue(
        "invalid_schema",
        path,
        keyword,
        "numeric schema keywords require type number or integer",
      );
    }
  }
  if (root && "enum" in node) {
    return memberIssue(
      "invalid_schema",
      path,
      "enum",
      "schema root enum is not supported",
    );
  }
  if ("pattern" in node) {
    if (typeof node.pattern !== "string") {
      return memberIssue(
        "invalid_schema",
        path,
        "pattern",
        "schema pattern must be a string",
      );
    }
    if (new TextEncoder().encode(node.pattern).byteLength > MAX_OUTPUT_PATTERN_BYTES) {
      return memberIssue(
        "limit_exceeded",
        path,
        "pattern",
        `schema pattern exceeds the maximum size of ${MAX_OUTPUT_PATTERN_BYTES} bytes`,
      );
    }
  }
  if ("enum" in node && (!Array.isArray(node.enum) || node.enum.length === 0)) {
    return memberIssue(
      "invalid_schema",
      path,
      "enum",
      "schema enum must be a nonempty array",
    );
  }
  const hasProperties = "properties" in node;
  const hasRequired = "required" in node;
  const hasAdditional = "additionalProperties" in node;
  if ((hasProperties || hasRequired || hasAdditional) && typeName !== "object") {
    const keyword = firstPresent(
      node,
      "additionalProperties",
      "properties",
      "required",
    )!;
    return memberIssue(
      "invalid_schema",
      path,
      keyword,
      "object schema keywords require type object",
    );
  }
  const propertyNames = new Set<string>();
  if (hasProperties) {
    if (!isObject(node.properties)) {
      return memberIssue(
        "invalid_schema",
        path,
        "properties",
        "schema properties must be an object",
      );
    }
    for (const name of Object.keys(node.properties).sort()) {
      const child = node.properties[name];
      const childPath = pointer(pointer(path, "properties"), name);
      if (!isObject(child) || Object.keys(child).length === 0) {
        return issue(
          "invalid_schema",
          childPath,
          undefined,
          `property ${JSON.stringify(name)} must contain a schema object`,
        );
      }
      propertyNames.add(name);
      const childIssue = validateSchemaNode(child, childPath, depth + 1, false);
      if (childIssue) return childIssue;
    }
  }
  if (hasRequired) {
    if (!Array.isArray(node.required)) {
      return memberIssue(
        "invalid_schema",
        path,
        "required",
        "schema required must be an array of property names",
      );
    }
    const seen = new Set<string>();
    for (const [index, value] of node.required.entries()) {
      const itemPath = pointer(pointer(path, "required"), String(index));
      if (typeof value !== "string" || value.length === 0) {
        return issue(
          "invalid_schema",
          itemPath,
          "required",
          "schema required must contain nonempty strings",
        );
      }
      if (seen.has(value)) {
        return issue(
          "invalid_schema",
          itemPath,
          "required",
          "schema required must not contain duplicates",
        );
      }
      if (!propertyNames.has(value)) {
        return issue(
          "invalid_schema",
          itemPath,
          "required",
          `required property ${JSON.stringify(value)} is not declared`,
        );
      }
      seen.add(value);
    }
  }
  if (hasAdditional) {
    const additional = node.additionalProperties;
    if (typeof additional !== "boolean") {
      const childPath = pointer(path, "additionalProperties");
      if (!isObject(additional)) {
        return memberIssue(
          "invalid_schema",
          path,
          "additionalProperties",
          "additionalProperties must be a boolean or schema object",
        );
      }
      if (Object.keys(additional).length === 0) {
        return issue(
          "invalid_schema",
          childPath,
          "additionalProperties",
          "additionalProperties must contain a schema object",
        );
      }
      const childIssue = validateSchemaNode(additional, childPath, depth + 1, false);
      if (childIssue) return childIssue;
    }
  }
  const hasItems = "items" in node;
  if (typeName === "array" && !hasItems) {
    return memberIssue("invalid_schema", path, "items", "array schemas require items");
  }
  if (hasItems) {
    if (typeName !== "array") {
      return memberIssue(
        "invalid_schema",
        path,
        "items",
        "schema items requires type array",
      );
    }
    if (!isObject(node.items) || Object.keys(node.items).length === 0) {
      return memberIssue(
        "invalid_schema",
        path,
        "items",
        "schema items must be a schema object",
      );
    }
    const childIssue = validateSchemaNode(
      node.items,
      pointer(path, "items"),
      depth + 1,
      false,
    );
    if (childIssue) return childIssue;
  }
  return undefined;
}

function validateBoundOrder(
  node: Record<string, unknown>,
  path: string,
): SchemaIssue | undefined {
  if (compareNumbers(node.minLength, node.maxLength) > 0) {
    return memberIssue(
      "invalid_schema",
      path,
      "maxLength",
      "schema minLength must not exceed maxLength",
    );
  }
  if (compareNumbers(node.minItems, node.maxItems) > 0) {
    return memberIssue(
      "invalid_schema",
      path,
      "maxItems",
      "schema minItems must not exceed maxItems",
    );
  }
  if (compareNumbers(node.minimum, node.maximum) > 0) {
    return memberIssue(
      "invalid_schema",
      path,
      "maximum",
      "schema minimum must not exceed maximum",
    );
  }
  return undefined;
}

function compareNumbers(left: unknown, right: unknown): number {
  if (!jsonNumber(left) || !jsonNumber(right)) return 0;
  return left - right;
}

function jsonNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function nonnegativeInteger(value: unknown): boolean {
  return jsonNumber(value) && Number.isInteger(value) && value >= 0;
}

function supportedType(value: string): boolean {
  return ["object", "array", "string", "number", "integer", "boolean"].includes(value);
}

function firstPresent(
  node: Record<string, unknown>,
  ...keywords: string[]
): string | undefined {
  return keywords.find((keyword) => keyword in node);
}

function isObject(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function memberIssue(
  code: SchemaIssueCode,
  path: string,
  keyword: string,
  message: string,
): SchemaIssue {
  return issue(code, pointer(path, keyword), keyword, message);
}

function issue(
  code: SchemaIssueCode,
  path: string,
  keyword: string | undefined,
  message: string,
): SchemaIssue {
  return { code, path, keyword, message };
}

function pointer(path: string, member: string): string {
  return `${path}/${member.replaceAll("~", "~0").replaceAll("/", "~1")}`;
}

File: sdk/rust/src/schema_preflight.rs
use std::collections::{HashMap, HashSet};

use serde_json::{json, Map, Value};

use crate::client::{ErrorCategory, NvokenError};

pub const MAX_OUTPUT_SCHEMA_BYTES: usize = 32 * 1024;
pub const MAX_OUTPUT_SCHEMA_DEPTH: usize = 16;
pub const MAX_OUTPUT_PATTERN_BYTES: usize = 1024;
pub const SCHEMA_PREFLIGHT_CODE: &str = "schema_preflight_failed";

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SchemaIssue {
    pub code: String,
    pub path: String,
    pub keyword: Option<String>,
    pub message: String,
}

pub fn preflight_output_schema(schema: &HashMap<String, Value>) -> Result<(), NvokenError> {
    let compact = serde_json::to_vec(schema).map_err(|_| {
        schema_error(issue(
            "invalid_schema",
            "",
            None,
            "schema must be JSON serializable",
        ))
    })?;
    if compact.len() > MAX_OUTPUT_SCHEMA_BYTES {
        return Err(schema_error(issue(
            "limit_exceeded",
            "",
            None,
            format!("compact schema exceeds {MAX_OUTPUT_SCHEMA_BYTES} bytes"),
        )));
    }
    let root: Map<String, Value> = schema
        .iter()
        .map(|(key, value)| (key.clone(), value.clone()))
        .collect();
    if let Some(issue) = validate_schema_node(&root, "", 1, true) {
        return Err(schema_error(issue));
    }
    Ok(())
}

fn schema_error(issue: SchemaIssue) -> NvokenError {
    let mut details = json!({
        "kind": "structured_output_schema",
        "code": issue.code,
        "path": issue.path,
    });
    if let Some(keyword) = &issue.keyword {
        details["keyword"] = Value::String(keyword.clone());
    }
    NvokenError {
        category: ErrorCategory::Validation,
        message: format!("output schema is invalid: {}", issue.message),
        status: None,
        code: Some(SCHEMA_PREFLIGHT_CODE.to_owned()),
        request_id: None,
        retry_after: None,
        details: Some(details),
    }
}

fn validate_schema_node(
    node: &Map<String, Value>,
    path: &str,
    depth: usize,
    root: bool,
) -> Option<SchemaIssue> {
    if depth > MAX_OUTPUT_SCHEMA_DEPTH {
        return Some(issue(
            "limit_exceeded",
            path,
            None,
            format!("schema exceeds the maximum nesting depth of {MAX_OUTPUT_SCHEMA_DEPTH}"),
        ));
    }
    let allowed = allowed_keywords();
    let mut keywords = node.keys().collect::<Vec<_>>();
    keywords.sort();
    for keyword in keywords {
        if !allowed.contains(keyword.as_str()) {
            return Some(member_issue(
                "unsupported_keyword",
                path,
                keyword,
                format!("unsupported schema keyword {keyword:?}"),
            ));
        }
    }
    let type_name = node.get("type").and_then(Value::as_str);
    if !type_name.is_some_and(supported_type) {
        return Some(member_issue(
            "invalid_schema",
            path,
            "type",
            "every schema position requires one supported string type",
        ));
    }
    let type_name = type_name.unwrap_or_default();
    if root && type_name != "object" {
        return Some(member_issue(
            "invalid_schema",
            path,
            "type",
            "schema root type must be object",
        ));
    }
    for keyword in ["description", "title"] {
        if node.get(keyword).is_some_and(|value| !value.is_string()) {
            return Some(member_issue(
                "invalid_schema",
                path,
                keyword,
                format!("schema {keyword} must be a string"),
            ));
        }
    }
    for keyword in ["maxItems", "maxLength", "minItems", "minLength"] {
        if node
            .get(keyword)
            .is_some_and(|value| !nonnegative_integer(value))
        {
            return Some(member_issue(
                "invalid_schema",
                path,
                keyword,
                format!("schema {keyword} must be a nonnegative integer"),
            ));
        }
    }
    for keyword in ["maximum", "minimum"] {
        if node.get(keyword).is_some_and(|value| !value.is_number()) {
            return Some(member_issue(
                "invalid_schema",
                path,
                keyword,
                format!("schema {keyword} must be a number"),
            ));
        }
    }
    if node
        .get("uniqueItems")
        .is_some_and(|value| !value.is_boolean())
    {
        return Some(member_issue(
            "invalid_schema",
            path,
            "uniqueItems",
            "schema uniqueItems must be a boolean",
        ));
    }
    if let Some(bound_issue) = validate_bound_order(node, path) {
        return Some(bound_issue);
    }
    if type_name != "string" {
        if let Some(keyword) = first_present(node, &["maxLength", "minLength", "pattern"]) {
            return Some(member_issue(
                "invalid_schema",
                path,
                keyword,
                "string schema keywords require type string",
            ));
        }
    }
    if type_name != "array" {
        if let Some(keyword) = first_present(node, &["maxItems", "minItems", "uniqueItems"]) {
            return Some(member_issue(
                "invalid_schema",
                path,
                keyword,
                "array schema bounds require type array",
            ));
        }
    }
    if type_name != "number" && type_name != "integer" {
        if let Some(keyword) = first_present(node, &["maximum", "minimum"]) {
            return Some(member_issue(
                "invalid_schema",
                path,
                keyword,
                "numeric schema keywords require type number or integer",
            ));
        }
    }
    if root && node.contains_key("enum") {
        return Some(member_issue(
            "invalid_schema",
            path,
            "enum",
            "schema root enum is not supported",
        ));
    }
    if let Some(pattern) = node.get("pattern") {
        let Some(pattern) = pattern.as_str() else {
            return Some(member_issue(
                "invalid_schema",
                path,
                "pattern",
                "schema pattern must be a string",
            ));
        };
        if pattern.len() > MAX_OUTPUT_PATTERN_BYTES {
            return Some(member_issue(
                "limit_exceeded",
                path,
                "pattern",
                format!(
                    "schema pattern exceeds the maximum size of {MAX_OUTPUT_PATTERN_BYTES} bytes"
                ),
            ));
        }
    }
    if node
        .get("enum")
        .is_some_and(|value| !value.as_array().is_some_and(|items| !items.is_empty()))
    {
        return Some(member_issue(
            "invalid_schema",
            path,
            "enum",
            "schema enum must be a nonempty array",
        ));
    }
    let has_properties = node.contains_key("properties");
    let has_required = node.contains_key("required");
    let has_additional = node.contains_key("additionalProperties");
    if (has_properties || has_required || has_additional) && type_name != "object" {
        let keyword = first_present(node, &["additionalProperties", "properties", "required"])
            .unwrap_or("properties");
        return Some(member_issue(
            "invalid_schema",
            path,
            keyword,
            "object schema keywords require type object",
        ));
    }
    let mut property_names = HashSet::new();
    if has_properties {
        let Some(properties) = node.get("properties").and_then(Value::as_object) else {
            return Some(member_issue(
                "invalid_schema",
                path,
                "properties",
                "schema properties must be an object",
            ));
        };
        let mut names = properties.keys().collect::<Vec<_>>();
        names.sort();
        for name in names {
            let child_path = pointer(&pointer(path, "properties"), name);
            let Some(child) = properties.get(name).and_then(Value::as_object) else {
                return Some(issue(
                    "invalid_schema",
                    &child_path,
                    None,
                    format!("property {name:?} must contain a schema object"),
                ));
            };
            if child.is_empty() {
                return Some(issue(
                    "invalid_schema",
                    &child_path,
                    None,
                    format!("property {name:?} must contain a schema object"),
                ));
            }
            property_names.insert(name.as_str());
            if let Some(child_issue) = validate_schema_node(child, &child_path, depth + 1, false) {
                return Some(child_issue);
            }
        }
    }
    if has_required {
        let Some(required) = node.get("required").and_then(Value::as_array) else {
            return Some(member_issue(
                "invalid_schema",
                path,
                "required",
                "schema required must be an array of property names",
            ));
        };
        let mut seen = HashSet::new();
        for (index, value) in required.iter().enumerate() {
            let item_path = pointer(&pointer(path, "required"), &index.to_string());
            let Some(name) = value.as_str().filter(|name| !name.is_empty()) else {
                return Some(issue(
                    "invalid_schema",
                    &item_path,
                    Some("required"),
                    "schema required must contain nonempty strings",
                ));
            };
            if !seen.insert(name) {
                return Some(issue(
                    "invalid_schema",
                    &item_path,
                    Some("required"),
                    "schema required must not contain duplicates",
                ));
            }
            if !property_names.contains(name) {
                return Some(issue(
                    "invalid_schema",
                    &item_path,
                    Some("required"),
                    format!("required property {name:?} is not declared"),
                ));
            }
        }
    }
    if has_additional {
        let additional = node.get("additionalProperties").unwrap_or(&Value::Null);
        if !additional.is_boolean() {
            let child_path = pointer(path, "additionalProperties");
            let Some(child) = additional.as_object() else {
                return Some(member_issue(
                    "invalid_schema",
                    path,
                    "additionalProperties",
                    "additionalProperties must be a boolean or schema object",
                ));
            };
            if child.is_empty() {
                return Some(issue(
                    "invalid_schema",
                    &child_path,
                    Some("additionalProperties"),
                    "additionalProperties must contain a schema object",
                ));
            }
            if let Some(child_issue) = validate_schema_node(child, &child_path, depth + 1, false) {
                return Some(child_issue);
            }
        }
    }
    let has_items = node.contains_key("items");
    if type_name == "array" && !has_items {
        return Some(member_issue(
            "invalid_schema",
            path,
            "items",
            "array schemas require items",
        ));
    }
    if has_items {
        if type_name != "array" {
            return Some(member_issue(
                "invalid_schema",
                path,
                "items",
                "schema items requires type array",
            ));
        }
        let Some(items) = node.get("items").and_then(Value::as_object) else {
            return Some(member_issue(
                "invalid_schema",
                path,
                "items",
                "schema items must be a schema object",
            ));
        };
        if items.is_empty() {
            return Some(member_issue(
                "invalid_schema",
                path,
                "items",
                "schema items must be a schema object",
            ));
        }
        if let Some(child_issue) =
            validate_schema_node(items, &pointer(path, "items"), depth + 1, false)
        {
            return Some(child_issue);
        }
    }
    None
}

fn validate_bound_order(node: &Map<String, Value>, path: &str) -> Option<SchemaIssue> {
    for (lower, upper, message) in [
        (
            "minLength",
            "maxLength",
            "schema minLength must not exceed maxLength",
        ),
        (
            "minItems",
            "maxItems",
            "schema minItems must not exceed maxItems",
        ),
        (
            "minimum",
            "maximum",
            "schema minimum must not exceed maximum",
        ),
    ] {
        let lower = node.get(lower).and_then(Value::as_f64);
        let upper_value = node.get(upper).and_then(Value::as_f64);
        if lower
            .zip(upper_value)
            .is_some_and(|(left, right)| left > right)
        {
            return Some(member_issue("invalid_schema", path, upper, message));
        }
    }
    None
}

// A bound is an integer when its value is one, not when its spelling is.
// `160`, `160.0`, and `1.6e2` are the same JSON number, and canonical JSON --
// which the runtime applies to tool input schemas before compiling them --
// renders every number in the last of those forms.
fn nonnegative_integer(value: &Value) -> bool {
    if value.as_u64().is_some() {
        return true;
    }
    value
        .as_f64()
        .is_some_and(|number| number.is_finite() && number >= 0.0 && number.fract() == 0.0)
}

fn first_present<'a>(node: &Map<String, Value>, keywords: &'a [&'a str]) -> Option<&'a str> {
    keywords
        .iter()
        .copied()
        .find(|keyword| node.contains_key(*keyword))
}

fn supported_type(value: &str) -> bool {
    matches!(
        value,
        "object" | "array" | "string" | "number" | "integer" | "boolean"
    )
}

fn allowed_keywords() -> HashSet<&'static str> {
    HashSet::from([
        "type",
        "title",
        "description",
        "properties",
        "required",
        "additionalProperties",
        "items",
        "enum",
        "pattern",
        "minLength",
        "maxLength",
        "minItems",
        "maxItems",
        "uniqueItems",
        "minimum",
        "maximum",
    ])
}

fn member_issue(code: &str, path: &str, keyword: &str, message: impl Into<String>) -> SchemaIssue {
    issue(code, &pointer(path, keyword), Some(keyword), message)
}

fn issue(code: &str, path: &str, keyword: Option<&str>, message: impl Into<String>) -> SchemaIssue {
    SchemaIssue {
        code: code.to_owned(),
        path: path.to_owned(),
        keyword: keyword.map(str::to_owned),
        message: message.into(),
    }
}

fn pointer(path: &str, member: &str) -> String {
    let escaped = member.replace('~', "~0").replace('/', "~1");
    format!("{path}/{escaped}")
}
